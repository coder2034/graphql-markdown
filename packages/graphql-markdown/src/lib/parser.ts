import {
  ASTNode,
  ArgumentNode,
  BooleanValueNode,
  DirectiveDefinitionNode,
  DirectiveNode,
  DocumentNode,
  EnumValueDefinitionNode,
  EnumValueNode,
  FieldDefinitionNode,
  FloatValueNode,
  InputValueDefinitionNode,
  IntValueNode,
  ListTypeNode,
  ListValueNode,
  NameNode,
  NamedTypeNode,
  NonNullTypeNode,
  ObjectFieldNode,
  ObjectTypeDefinitionNode,
  ObjectValueNode,
  OperationTypeDefinitionNode,
  OperationTypeNode,
  StringValueNode,
  TypeDefinitionNode,
  visit,
} from "graphql";

import { ParsedNode } from "..";

const GRAPHQL_KIND = {
  DirectiveDefinition: "DirectiveDefinition",
  EnumTypeDefinition: "EnumTypeDefinition",
  InputObjectTypeDefinition: "InputObjectTypeDefinition",
  InputTypeDefinition: "InputTypeDefinition",
  InterfaceTypeDefinition: "InterfaceTypeDefinition",
  ObjectTypeDefinition: "ObjectTypeDefinition",
  OperationTypeDefinition: "OperationTypeDefinition",
  ScalarTypeDefinition: "ScalarTypeDefinition",
  UnionTypeDefinition: "UnionTypeDefinition",
} as const;

const SIMPLIFIED_NODE_KIND = {
  DirectiveDefinition: "directive",
  EnumTypeDefinition: "enum",
  InputObjectTypeDefinition: "input",
  InputTypeDefinition: "interface",
  Mutation: "mutation",
  ObjectTypeDefinition: "object",
  OperationTypeDefinition: "operation",
  Query: "query",
  ScalarTypeDefinition: "scalar",
  Subscription: "subscription",
  UnionTypeDefinition: "union",
} as const;

const OPERATION_TYPE = {
  MUTATION: "Mutation",
  QUERY: "Query",
  SUBSCRIPTION: "Subscription",
} as const;

const visitor = {
  Argument: (node: ArgumentNode) => {
    return node;
  },
  BooleanValue: ({ value }: BooleanValueNode) => {
    return value;
  },
  Directive: (node: DirectiveNode) => {
    return { ...node, name: `@${node.name}` };
  },
  DirectiveDefinition: (node: DirectiveDefinitionNode) => {
    return { ...node, name: `@${node.name}` };
  },
  Document: ({ definitions }: DocumentNode) => {
    return definitions;
  },
  EnumValue: ({ value }: EnumValueNode) => {
    return value;
  },
  EnumValueDefinition: (node: EnumValueDefinitionNode) => {
    return node;
  },
  FieldDefinition: ({
    type,
    name,
    arguments: args,
    directives,
    kind,
    description,
  }: FieldDefinitionNode) => {
    return {
      ...type,
      arguments: args,
      description,
      directives,
      kind,
      name,
    };
  },
  FloatValue: ({ value }: FloatValueNode) => {
    return value;
  },
  InputValueDefinition: ({
    type,
    name,
    directives,
    defaultValue,
    kind,
    description,
  }: InputValueDefinitionNode) => {
    return {
      ...type,
      defaultValue,
      description,
      directives,
      kind,
      name,
    };
  },
  IntValue: ({ value }: IntValueNode) => {
    return value;
  },
  ListType: ({ type }: ListTypeNode) => {
    return {
      ...type,
      isList: true,
    };
  },
  ListValue: ({ values }: ListValueNode) => {
    return values;
  },
  Name: ({ value }: NameNode) => {
    return value;
  },
  NamedType: ({ name }: NamedTypeNode) => {
    return {
      isList: false,
      isNull: true,
      type: name,
    };
  },
  NonNullType: ({ type }: NonNullTypeNode) => {
    return {
      ...type,
      isNull: false,
    };
  },
  NullValue: () => {
    return "null";
  },
  ObjectField: (node: ObjectFieldNode) => {
    return node;
  },
  ObjectValue: ({ fields }: ObjectValueNode) => {
    return fields;
  },
  OperationType: (node: OperationTypeNode) => {
    return node.toString();
  },
  OperationTypeDefinition: (node: OperationTypeDefinitionNode) => {
    return node;
  },
  StringValue: ({ value }: StringValueNode) => {
    return value;
  },
  TypeDefinition: (node: TypeDefinitionNode) => {
    return node;
  },
} as const;

const OperationTypes: readonly string[] = [
  OPERATION_TYPE.QUERY,
  OPERATION_TYPE.MUTATION,
  OPERATION_TYPE.SUBSCRIPTION,
] as const;

const isNodeTypeOperation = (
  node: ASTNode | ParsedNode
): node is ObjectTypeDefinitionNode => {
  return (
    "name" in node &&
    typeof node.name !== "undefined" &&
    OperationTypes.includes(node.name.toString()) &&
    "fields" in node
  );
};

export const getSimplifiedNodeKind = (node: ASTNode | ParsedNode): string => {
  const kind = "operation" in node ? node.operation : node.kind;
  return SIMPLIFIED_NODE_KIND[kind] ?? node.kind;
};

export const parseSchema = (schema: DocumentNode): ParsedNode[] => {
  const nodes: ASTNode[] = visit(schema, { leave: visitor });
  return nodes.flatMap((node) => {
    if (isNodeTypeOperation(node) && typeof node.fields !== "undefined") {
      return node.fields.map<ParsedNode>((operation: FieldDefinitionNode) => {
        const parsedNode = {
          ...operation,
          kind: GRAPHQL_KIND.OperationTypeDefinition,
          operation: node.name,
        } as unknown as ParsedNode;
        return {
          ...parsedNode,
          simplifiedKind: getSimplifiedNodeKind(parsedNode),
        } as ParsedNode;
      });
    }
    return {
      ...node,
      simplifiedKind: getSimplifiedNodeKind(node),
    } as ParsedNode;
  });
};
