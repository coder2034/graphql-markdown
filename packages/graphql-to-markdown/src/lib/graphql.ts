import {
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLScalarType,
  isListType,
  GraphQLID,
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLInputObjectType,
  GraphQLFieldMap,
  GraphQLField,
  GraphQLSchema,
  GraphQLNamedType,
} from "graphql";

import { Maybe } from "graphql/jsutils/Maybe";
import { TypeMap } from "graphql/type/schema";

import { hasOwnMethod, hasOwnProperty } from ".";

export {
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLScalarType,
  isListType,
  GraphQLID,
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLInputObjectType,
  isDirective as isDirectiveType,
  getNamedType,
  isScalarType,
  isEnumType,
  isUnionType,
  isInterfaceType,
  isObjectType,
  isInputObjectType as isInputType,
  isNullableType,
  printSchema,
} from "graphql";
export { loadSchema } from "@graphql-tools/load";
export { GraphQLFileLoader } from "@graphql-tools/graphql-file-loader";
export { UrlLoader } from "@graphql-tools/url-loader";
export { JsonFileLoader } from "@graphql-tools/json-file-loader";

export const SCHEMA_EXCLUDE_LIST_PATTERN =
  /^(?!Query$|Mutation$|Subscription$|__.+$).*$/;

export function getDefaultValue(argument: { type: unknown; defaultValue: unknown }) {
  if (isListType(argument.type)) {
    return `[${argument.defaultValue || ""}]`;
  }

  switch (argument.type) {
    case GraphQLID:
    case GraphQLInt:
      return `${argument.defaultValue || "0"}`;
    case GraphQLFloat:
      return `${argument.defaultValue || "0.0"}`;
    case GraphQLString:
    default:
      return argument.defaultValue ? `"${argument.defaultValue}"` : undefined;
  }
}

export function getFilteredTypeMap(
  typeMap: Record<string, GraphQLNamedType>,
  excludeList = SCHEMA_EXCLUDE_LIST_PATTERN,
): Maybe<TypeMap> {
  if (!typeMap) return;
  return Object.keys(typeMap)
    .filter((key) => excludeList.test(key))
    .reduce((res: TypeMap, key) => ((res[key] = typeMap[key]), res), {});
}

export function getIntrospectionFieldsList(queryType: Maybe<GraphQLObjectType<unknown, unknown>>): Maybe<GraphQLFieldMap<unknown, unknown>> {
  if (!queryType){
    return;
  }

  if (!hasOwnMethod(queryType, "getFields")) {
    return;
  }
  return queryType.getFields();
}

export function getFields(type: GraphQLObjectType<unknown, unknown>): Maybe<GraphQLField<any, unknown, Record<string, unknown>>[]> {
  if (!hasOwnMethod(type, "getFields")) {
    return;
  }
  const fieldMap = type.getFields();
  return Object.keys(fieldMap).map((name) => fieldMap[name]);
}

export function getTypeName(
  type: GraphQLObjectType<unknown, unknown>,
  defaultName = "",
): Maybe<string> {
  if (!type) {
    return;
  }
  return (
    (hasOwnProperty(type, "name") && type.name) ||
    (hasOwnMethod(type, "toString") && type.toString()) ||
    defaultName
  );
}

export function getTypeFromTypeMap(typeMap: TypeMap, type: any): Maybe<Record<string, unknown>> {
  if (!typeMap) return;
  return Object.keys(typeMap)
    .filter((key) => typeMap[key] instanceof type)
    .reduce((res: any, key) => ((res[key] = typeMap[key]), res), {});
}

export function getSchemaMap(schema: GraphQLSchema) {
  const typeMap = getFilteredTypeMap(schema.getTypeMap());
  if (!typeMap) return;
  return {
    queries: getIntrospectionFieldsList(
      schema.getQueryType && schema.getQueryType(),
    ),
    mutations: getIntrospectionFieldsList(
      schema.getMutationType && schema.getMutationType(),
    ),
    subscriptions: getIntrospectionFieldsList(
      schema.getSubscriptionType && schema.getSubscriptionType(),
    ),
    directives: schema.getDirectives(),
    objects: getTypeFromTypeMap(typeMap, GraphQLObjectType),
    unions: getTypeFromTypeMap(typeMap, GraphQLUnionType),
    interfaces: getTypeFromTypeMap(typeMap, GraphQLInterfaceType),
    enums: getTypeFromTypeMap(typeMap, GraphQLEnumType),
    inputs: getTypeFromTypeMap(typeMap, GraphQLInputObjectType),
    scalars: getTypeFromTypeMap(typeMap, GraphQLScalarType),
  };
}

export function isParametrizedField(field: GraphQLFieldMap<unknown, unknown> | GraphQLField<unknown, unknown, unknown>) {
  return hasOwnProperty(field, "args") && field.args.length > 0;
}

export function isOperation(query: GraphQLFieldMap<unknown, unknown> | GraphQLField<unknown, unknown, unknown>) {
  return hasOwnProperty(query, "type");
}
