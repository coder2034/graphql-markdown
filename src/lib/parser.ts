import {
  ArgumentNode,
  BooleanValueNode,
  DirectiveDefinitionNode,
  DocumentNode,
  EnumValueDefinitionNode,
  EnumValueNode,
  FieldDefinitionNode,
  FloatValueNode,
  InputValueDefinitionNode,
  IntValueNode,
  ListTypeNode,
  ListValueNode,
  NameNode,
  NamedTypeNode,
  NonNullTypeNode,
  ObjectFieldNode,
  ObjectValueNode,
  OperationTypeDefinitionNode,
  OperationTypeNode,
  StringValueNode,
  TypeDefinitionNode,
  visit,
} from "graphql";

const visitor = {
  Argument: (node: ArgumentNode) => {
    return node;
  },
  BooleanValue: ({ value }: BooleanValueNode) => {
    return value;
  },
  Document: ({ definitions }: DocumentNode) => {
    return definitions;
  },
  EnumValue: ({ value }: EnumValueNode) => {
    return value;
  },
  EnumValueDefinition: (node: EnumValueDefinitionNode) => {
    return node;
  },
  FieldDefinition: ({
    type,
    name,
    arguments: args,
    directives,
    kind,
    description,
  }: FieldDefinitionNode) => {
    return {
      ...type,
      arguments: args,
      description,
      directives,
      kind,
      name,
    };
  },
  FloatValue: ({ value }: FloatValueNode) => {
    return value;
  },
  InputValueDefinition: ({
    type,
    name,
    directives,
    defaultValue,
    kind,
    description,
  }: InputValueDefinitionNode) => {
    return {
      ...type,
      defaultValue,
      description,
      directives,
      kind,
      name,
    };
  },
  IntValue: ({ value }: IntValueNode) => {
    return value;
  },
  ListType: ({ type }: ListTypeNode) => {
    return {
      ...type,
      isList: true,
    };
  },
  ListValue: ({ values }: ListValueNode) => {
    return values;
  },
  Name: ({ value }: NameNode) => {
    return value;
  },
  NamedType: ({ name }: NamedTypeNode) => {
    return {
      isList: false,
      isNull: true,
      type: name,
    };
  },
  NonNullType: ({ type }: NonNullTypeNode) => {
    return {
      ...type,
      isNull: false,
    };
  },
  NullValue: () => {
    return "null";
  },
  ObjectField: (node: ObjectFieldNode) => {
    return node;
  },
  ObjectValue: ({ fields }: ObjectValueNode) => {
    return fields;
  },
  OperationType: (node: OperationTypeNode) => {
    return node.toString();
  },
  OperationTypeDefinition: (node: OperationTypeDefinitionNode) => {
    return node;
  },
  StringValue: ({ value }: StringValueNode) => {
    return value;
  },
  TypeDefinition: (node: TypeDefinitionNode) => {
    return node;
  },
  TypeSystemDefinition: (
    node: DirectiveDefinitionNode | TypeDefinitionNode
  ) => {
    return node;
  },
};

export const parseSchema = (schema: DocumentNode): any => {
  return visit(schema, { leave: visitor });
};
