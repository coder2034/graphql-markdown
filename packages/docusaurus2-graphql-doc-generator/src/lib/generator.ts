import { promises as fs } from "fs";
import path from "path";

import prettier from "prettier";

import { generateMarkdownFromSchema } from "@edno/graphql-markdown";

import { PluginOptions, Sidebar } from "../types";

const SIDEBAR_FILE = "sidebar-schema.js";

const prettifyJSON = (content: string): string => {
  return prettier.format(content, { parser: "json" });
};

const prettifyMarkdown = (content: string): string => {
  return prettier.format(content, { parser: "markdown" });
};

const renderHomepage = async (
  homepageLocation: string,
  options: PluginOptions
): Promise<string> => {
  const homepage = await fs.readFile(homepageLocation, "utf8");

  const filePath = path.join(options.rootPath, path.basename(homepageLocation));
  await fs.writeFile(filePath, prettifyMarkdown(homepage), "utf8");

  return path.relative("./", filePath);
};

const renderSidebar = async ({
  rootPath,
  baseURL,
}: PluginOptions): Promise<string> => {
  const sidebar: Record<string, Sidebar> = {
    schemaSidebar: [
      {
        dirName: baseURL,
        type: "autogenerated",
      },
    ],
  };

  const filePath = path.join(rootPath, SIDEBAR_FILE);
  await fs.writeFile(filePath, prettifyJSON(JSON.stringify(sidebar)), "utf8");

  return path.relative("./", filePath);
};

export const generateDocFromSchema = async (
  schemaLocation: string,
  options: PluginOptions
): Promise<{ pages: number; sidebar: string }> => {
  const pages = await generateMarkdownFromSchema({ saveToFiles: true });

  // eslint-disable-next-line no-magic-numbers
  if (typeof pages === "undefined" || pages.length === 0) {
    throw new Error("Parsing error, no pages generated");
  }

  if (typeof options === "undefined") {
    throw new Error("Parsing error, no pages generated");
  }

  await renderHomepage(options.homepage, options);
  const sidebarPath = await renderSidebar(options);

  return { pages: pages.length, sidebar: sidebarPath };
};
