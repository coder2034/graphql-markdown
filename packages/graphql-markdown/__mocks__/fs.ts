import { IFS } from "unionfs/lib/fs";
import { Volume } from "memfs";
import { ufs } from "unionfs";

const fs = jest.requireActual("fs");

const vol = Volume.fromJSON({});
vol.mkdirSync(process.cwd(), { recursive: true });

module.exports = ufs.use(fs).use(vol as unknown as IFS);
