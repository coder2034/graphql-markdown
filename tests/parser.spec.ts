import { parseSchema } from "../src/lib/parser";

import { promises as fs } from "fs";
import * as path from "path";

import { DocumentNode } from "graphql";

describe("parseSchema", () => {
  let schema: DocumentNode;

  beforeAll(async () => {
    const schemaLocation = path.resolve(
      `${__dirname}/__data__/schema/tweet.json`
    );
    schema = JSON.parse((
      await fs.readFile(schemaLocation)
    ).toString()) as unknown as DocumentNode;
  });

  it("should returned a parsed schema ", () => {
    const parsed = parseSchema(schema);
    expect(parsed).toMatchSnapshot();
  });
});
