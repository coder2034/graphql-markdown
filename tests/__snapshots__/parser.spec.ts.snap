// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseSchema should returned a parsed schema 1`] = `
Array [
  Object {
    "directives": Array [
      Object {
        "arguments": Array [
          Object {
            "kind": "Argument",
            "name": "url",
            "value": "https://www.w3.org/TR/SRI",
          },
        ],
        "kind": "Directive",
        "name": "@specifiedBy",
      },
    ],
    "kind": "ScalarTypeDefinition",
    "name": "SRI",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "id",
        "type": "ID",
      },
    ],
    "interfaces": Array [],
    "kind": "InterfaceTypeDefinition",
    "name": "Node",
  },
  Object {
    "directives": Array [],
    "kind": "EnumTypeDefinition",
    "name": "MediaType",
    "values": Array [
      Object {
        "directives": Array [],
        "kind": "EnumValueDefinition",
        "name": "IMAGE",
      },
      Object {
        "directives": Array [],
        "kind": "EnumValueDefinition",
        "name": "VIDEO",
      },
    ],
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "url",
        "type": "Url",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "sri",
        "type": "SRI",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "type",
        "type": "MediaType",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Media",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "message",
        "type": "String",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Error",
  },
  Object {
    "directives": Array [],
    "kind": "UnionTypeDefinition",
    "name": "TweetResponse",
    "types": Array [
      Object {
        "isList": false,
        "isNull": true,
        "type": "Tweet",
      },
      Object {
        "isList": false,
        "isNull": true,
        "type": "Error",
      },
    ],
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "id",
        "type": "ID",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "body",
        "type": "String",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "date",
        "type": "Date",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "Author",
        "type": "User",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "Stats",
        "type": "Stat",
      },
      Object {
        "arguments": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "isList": false,
            "isNull": true,
            "kind": "InputValueDefinition",
            "name": "meta",
            "type": "String",
          },
        ],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "getMeta",
        "type": "Meta",
      },
    ],
    "interfaces": Array [
      Object {
        "isList": false,
        "isNull": true,
        "type": "Node",
      },
    ],
    "kind": "ObjectTypeDefinition",
    "name": "Tweet",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "id",
        "type": "ID",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "username",
        "type": "String",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "first_name",
        "type": "String",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "last_name",
        "type": "String",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "full_name",
        "type": "String",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [
          Object {
            "arguments": Array [],
            "kind": "Directive",
            "name": "@deprecated",
          },
        ],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "name",
        "type": "String",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [
          Object {
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": "reason",
                "value": "Replaced by the \`Media\` field \`avatar\`.",
              },
            ],
            "kind": "Directive",
            "name": "@deprecated",
          },
        ],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "avatar_url",
        "type": "Url",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "avatar",
        "type": "Media",
      },
    ],
    "interfaces": Array [
      Object {
        "isList": false,
        "isNull": true,
        "type": "Node",
      },
    ],
    "kind": "ObjectTypeDefinition",
    "name": "User",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "views",
        "type": "Int",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [
          Object {
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": "reason",
                "value": "Use \`reactions\`.",
              },
            ],
            "kind": "Directive",
            "name": "@deprecated",
          },
        ],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "likes",
        "type": "Int",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "reactions",
        "type": "Int",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "retweets",
        "type": "Int",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "responses",
        "type": "Int",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Stat",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "id",
        "type": "ID",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "date",
        "type": "Date",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "type",
        "type": "String",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Notification",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "count",
        "type": "Int",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Meta",
  },
  Object {
    "directives": Array [],
    "kind": "ScalarTypeDefinition",
    "name": "Url",
  },
  Object {
    "directives": Array [],
    "kind": "ScalarTypeDefinition",
    "name": "Date",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "isList": false,
            "isNull": false,
            "kind": "InputValueDefinition",
            "name": "id",
            "type": "ID",
          },
        ],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "Tweet",
        "type": "TweetResponse",
      },
      Object {
        "arguments": Array [
          Object {
            "defaultValue": "5",
            "description": undefined,
            "directives": Array [],
            "isList": false,
            "isNull": true,
            "kind": "InputValueDefinition",
            "name": "limit",
            "type": "Int",
          },
          Object {
            "defaultValue": "0",
            "description": undefined,
            "directives": Array [],
            "isList": false,
            "isNull": true,
            "kind": "InputValueDefinition",
            "name": "skip",
            "type": "Int",
          },
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "isList": false,
            "isNull": true,
            "kind": "InputValueDefinition",
            "name": "sort_field",
            "type": "String",
          },
          Object {
            "defaultValue": "ASC",
            "description": undefined,
            "directives": Array [],
            "isList": false,
            "isNull": true,
            "kind": "InputValueDefinition",
            "name": "sort_order",
            "type": "String",
          },
        ],
        "description": undefined,
        "directives": Array [],
        "isList": true,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "Tweets",
        "type": "TweetResponse",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [
          Object {
            "arguments": Array [],
            "kind": "Directive",
            "name": "@deprecated",
          },
        ],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "TweetsMeta",
        "type": "Meta",
      },
      Object {
        "arguments": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "isList": false,
            "isNull": false,
            "kind": "InputValueDefinition",
            "name": "id",
            "type": "ID",
          },
        ],
        "description": undefined,
        "directives": Array [
          Object {
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": "reason",
                "value": "Use \`Users\`.",
              },
            ],
            "kind": "Directive",
            "name": "@deprecated",
          },
        ],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "User",
        "type": "User",
      },
      Object {
        "arguments": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "isList": true,
            "isNull": false,
            "kind": "InputValueDefinition",
            "name": "id",
            "type": "ID",
          },
        ],
        "description": undefined,
        "directives": Array [],
        "isList": true,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "Users",
        "type": "User",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [
          Object {
            "arguments": Array [],
            "kind": "Directive",
            "name": "@deprecated",
          },
        ],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "NotificationsMeta",
        "type": "Meta",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Query",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [
              Object {
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": "reason",
                    "value": "Use \`content\`.",
                  },
                ],
                "kind": "Directive",
                "name": "@deprecated",
              },
            ],
            "isList": false,
            "isNull": true,
            "kind": "InputValueDefinition",
            "name": "input",
            "type": "TweetContent",
          },
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "isList": false,
            "isNull": true,
            "kind": "InputValueDefinition",
            "name": "content",
            "type": "TweetContent",
          },
        ],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "createTweet",
        "type": "Tweet",
      },
      Object {
        "arguments": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "isList": false,
            "isNull": false,
            "kind": "InputValueDefinition",
            "name": "id",
            "type": "ID",
          },
        ],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "deleteTweet",
        "type": "Tweet",
      },
      Object {
        "arguments": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "isList": false,
            "isNull": false,
            "kind": "InputValueDefinition",
            "name": "id",
            "type": "ID",
          },
        ],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "markTweetRead",
        "type": "Boolean",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Mutation",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [
          Object {
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": "reason",
                "value": "Use \`body\`.",
              },
            ],
            "kind": "Directive",
            "name": "@deprecated",
          },
        ],
        "isList": false,
        "isNull": true,
        "kind": "InputValueDefinition",
        "name": "content",
        "type": "String",
      },
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "InputValueDefinition",
        "name": "body",
        "type": "String",
      },
    ],
    "kind": "InputObjectTypeDefinition",
    "name": "TweetContent",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "isList": false,
            "isNull": true,
            "kind": "InputValueDefinition",
            "name": "limit",
            "type": "Int",
          },
        ],
        "description": undefined,
        "directives": Array [],
        "isList": true,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "Notifications",
        "type": "Notification",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Subscription",
  },
]
`;
